sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -876,7 +885,30 @@ class Dataset[T] private[sql](
    * @since 2.0.0
    */
   def join(right: Dataset[_]): DataFrame = withPlan {
-    Join(logicalPlan, right.logicalPlan, joinType = Inner, None, JoinHint.NONE)
+    val (joinLeft, joinRight) = prepareJoinPlan(this, right)
+    Join(joinLeft, joinRight, joinType = Inner, None, JoinHint.NONE)
+  }
+
+  // Called by `Dataset#join`, to attach the Dataset id to the logical plan, so that we
+  // can resolve column reference correctly later. See `ResolveDatasetColumnReference`.
+  private def createPlanWithDatasetId(): LogicalPlan = {
+    if (!sparkSession.sessionState.conf.getConf(SQLConf.RESOLVE_DATASET_COLUMN_REFERENCE)) {
+      return logicalPlan
+    }
+
+    // The alias should start with `SubqueryAlias.HIDDEN_ALIAS_PREFIX`, so that `SubqueryAlias` can
+    // recognize it and keep the output qualifiers unchanged.
+    SubqueryAlias(s"${SubqueryAlias.HIDDEN_ALIAS_PREFIX}${Dataset.ID_PREFIX}_$id", logicalPlan)
```

Comment:
Here I am missing something. Sorry if the question is dumb, but I can't get why do we need to add this "special" node for it?

Reply:
to attach the dataset id to the logical plan, so that we can resolve column reference later.

Reply:
I'd rather add the `datasetId` info to the `LogicalPlan` and avoid introducing a new plan here. I think it would be easier in this way to generalize this approach to other cases when the same problem may arise. What do you think?

Reply:
How can we add dataset id to every kind of `LogicalPlan`?

Reply:
I think that could be an option. And for the moment we could add it only to the child/children of a join since we only need it there. But I see there is no guarantee that the plan(s) are not replaces/removed during the analysis/optimization phase, so it may not be doable indeed.